active/XamlServiceProvider/XamlServiceProviderApi.md

```diff
@@ -0,0 +1,251 @@
+
+When you write a custom markup extension (see the Description section below), you override the `ProvideValue` method to return the built value. The ProvideValue method is different in  Xaml (WinUI) than it is in WPF, because WPF passes an extra [IServiceProvider](http://msdn.microsoft.com/library/System.IServiceProvider) parameter:
+
+Xaml:
+
+```cs
+protected override object ProvideValue(IServiceProvider serviceProvider)
+{
+    return DateTime.Now.ToString(Format);
+}
+```
+
+WPF:
+
+```cs
+protected override object ProvideValue(IServiceProvider serviceProvider)
+{
+    return DateTime.Now.ToString(Format);
+}
+```
+
+The service provider can be used, for example, to discover the property being assigned to.
+
+This spec is adding a new version of ProvideValue that passes a service provider.
+
+> To do: This is defining an IXamlServiceProvider, can/should this project in .Net to [System.IServiceProvider](http://msdn.microsoft.com/library/System.IServiceProvider)?
+
+# Description
+
+A developer defines a custom markup extension by subclassing the MarkupExtension class and overriding the ProvideValue method. This can then be used in Xaml markup: the markup extension can be created using the "{ }" markup expression syntax, and the return value from ProvideValue will be set as the property value.
+
+For example, given this markup extension:
+
+```cs
+public class CurrentDate : MarkupExtension
+{
+    public string Format { get; set; } = "";
+    protected override object ProvideValue()
+    {
+        return DateTime.Now.ToString(Format);
+    }
+}
+```
+
+you can do this in Xaml markup:
+
+```xml
+<TextBlock>
+    Today is:  
+    <Run Text="{local:CurrentDate Format=d}"/>
+</TextBlock>
+```
+
+On an en-US machine, this will produce something like:
+ 
+![MarkupExtension sample](MESample.png)
+
+## MarkupExtension.ProvideValue virtual method
+
+A markup extension overrides the ProvideValue method, and returns a value from it that the Xaml loader sets as a property value.
+
+There are two versions of ProvideValue, one that passes no parameters and one that passes a service provider. From the service provider, a markup extension implementation can retrieve the following services:
+* `IProvideValueTarget`, which provides information about the type/property the markup extension is being applied to. This aligns with WPF's [IProvideValueTarget](https://docs.microsoft.com/dotnet/api/System.Windows.Markup.IProvideValueTarget).
+* `IRootObjectProvider`, which provides a reference to the root object in the markup. This aligns with WPF's [IRootObjectProvider](https://docs.microsoft.com/dotnet/api/System.Xaml.IRootObjectProvider).
+* `IUriContext` , which provides the base URI of the markup. This aligns with WPF's [IUriContext](https://docs.microsoft.com/dotnet/api/System.Windows.Markup.IUriContext).
+

```

Comment:
Was ['IXamlTypeResolver'](https://docs.microsoft.com/en-us/dotnet/api/system.windows.markup.ixamltyperesolver?view=netcore-3.1) inadvertently omitted from this list?